GenTs.cs

```diff
@@ -0,0 +1,297 @@
+          break;
+        default:
+          if (type == CiSystem.MatchClass)
+          {
+            Write("RegExp");
+          }
+          else
+          {
+            Write(type.Name);
+          }
+          break;
+      }
+    }
+
+    void Write(CiVisibility visibility)
+    {
+      switch (visibility)
+      {
+        case CiVisibility.Private:
+          Write("private ");
+          break;
+        case CiVisibility.Internal:
+          break;
+        case CiVisibility.Protected:
+          Write("protected ");
+          break;
+        case CiVisibility.Public:
+          Write("public ");
+          break;
+      }
+    }
+
+    void WriteSignature(CiMethod method, int paramCount)
+    {
+      WriteDoc(method);
+      Write(method.Visibility);
+      switch (method.CallType)
+      {
+        case CiCallType.Static:
+          Write("static ");
+          break;
+        case CiCallType.Virtual:
+          break;
+        case CiCallType.Abstract:
+          Write("abstract ");
+          break;
+        case CiCallType.Override:
+          break;
+        case CiCallType.Normal:
+          // no final methods in TS
+          break;
+        case CiCallType.Sealed:
+          // no final methods in TS
+          break;
+        default:
+          throw new NotImplementedException(method.CallType.ToString());
+      }
+      WriteName(method);
+      Write('(');
+      int i = 0;
+      foreach (CiVar param in method.Parameters)
+      {
+        if (i >= paramCount)
+          break;
+        if (i > 0)
+          Write(", ");
+        WriteName(param);
+        if (param.Value != null)
+          Write('?');
+        Write(": ");
+        Write(param.Type);
+        i++;
+      }
+      Write("): ");
+      Write(method.Type);
+      WriteLine(";");
+    }
+
+    void WriteConsts(IEnumerable<CiConst> consts)
+    {
+      foreach (CiConst konst in consts)
+      {
+        WriteLine();
+        Write(konst.Documentation);
+        Write(konst.Visibility);
+        Write("static readonly ");
+        WriteTypeAndName(konst);
+        WriteLine(';');
+      }
+    }
+
+    void Write(CiClass klass)
+    {
+      Write(klass.Documentation);
+      Write("export ");
+      switch (klass.CallType)
+      {
+        case CiCallType.Normal:
+          break;
+        case CiCallType.Abstract:
+          Write("abstract ");
+          break;
+        case CiCallType.Static:
+        case CiCallType.Sealed:
+          // there's no final/sealed keyword, but we can accomplish it by marking the constructor private
+          break;
+        default:
+          throw new NotImplementedException(klass.CallType.ToString());
+      }
+      OpenClass(klass, "", " extends ");
+
+      if (klass.CallType == CiCallType.Static || klass.CallType == CiCallType.Sealed)
+      {
+        Write("private constructor();");
+      }
+      else if (klass.Constructor != null && klass.Constructor.Visibility != CiVisibility.Public)
```

Comment:
Why emit only non-public constructors?

Reply:
Well, a public zero-argument constructor is implied if there's no constructor in a class declaration.  Wouldn't hurt to emit it anyway though

Reply:
How about a documentation comment for the public constructor?

Reply:
Yeah makes sense